src/Driver/PDO/Connection.php

```diff
@@ -88,17 +92,86 @@ public function quote(string $value): string
+    /**
+     * {@inheritDoc}
+     */
+    public function getLastInsertId()
     {
+        $driverName = $this->connection->getAttribute(PDO::ATTR_DRIVER_NAME);
+
+        if ($driverName === 'oci') {
+            throw IdentityColumnsNotSupported::new();
+        }
+
         try {
-            if ($name === null) {
-                return $this->connection->lastInsertId();
+            $lastInsertId = $this->connection->lastInsertId();
+        } catch (PDOException $exception) {
+            [$sqlState] = $exception->errorInfo;
+
+            // PDO PGSQL throws a 'lastval is not yet defined in this session' error when no identity value is
+            // available, with SQLSTATE 55000 'Object Not In Prerequisite State'
+            if ($sqlState === '55000' && $this->connection->getAttribute(PDO::ATTR_DRIVER_NAME) === 'pgsql') {
+                throw NoIdentityValue::new($exception);
             }
 
-            return $this->connection->lastInsertId($name);
+            throw Exception::new($exception);
+        }
+
+        if (! $this->isValidLastInsertIdOrSequenceNumber($lastInsertId)) {
+            throw NoIdentityValue::new();
+        }
+
+        return $lastInsertId;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getSequenceNumber(string $name)
+    {
+        $driverName = $this->connection->getAttribute(PDO::ATTR_DRIVER_NAME);
+
+        if ($driverName === 'mysql') {
+            throw SequencesNotSupported::new();
+        }
+
+        if ($driverName === 'sqlite') {
+            // SQLite does not support sequences. However, PDO::lastInsertId() ignores the name parameter, and returns
+            // the last insert ID even if a sequence name is given. We expect an exception in that case.
+            throw SequencesNotSupported::new();
+        }
+
+        if ($driverName === 'oci') {
+            throw new SequencesNotSupported('Oracle supports sequences, but the PDO OCI driver does not support them.');
+        }
+
+        try {
+            $sequenceNumber = $this->connection->lastInsertId($name);
         } catch (PDOException $exception) {
+            [$sqlState] = $exception->errorInfo;
+
+            // PDO PGSQL throws an 'Undefined table' when the sequence does not exist
+            if ($sqlState === '42P01' && $this->connection->getAttribute(PDO::ATTR_DRIVER_NAME) === 'pgsql') {
+                throw SequenceDoesNotExist::new($name, $exception);
+            }
+
             throw Exception::new($exception);
         }
+
+        if (! $this->isValidLastInsertIdOrSequenceNumber($sequenceNumber)) {
+            throw SequenceDoesNotExist::new($name);
+        }
+
+        return $sequenceNumber;
+    }
+
+    /**
+     * @param int|string $value
+     */
+    private function isValidLastInsertIdOrSequenceNumber($value): bool
+    {
+        // pdo_mysql & pdo_sqlite return 0 or '0', pdo_sqlsrv returns ''
+        return $value !== 0 && $value !== '0' && $value !== '';
```

Comment:
By the moment when this code is executed, we know the driver name. Should we make this code explicitly dependent on the driver?